File: app-ios/VelaWallet/VelaWallet/Components/StepProgress.swift
//
//  StepProgress.swift
//  VelaWallet
//
//  The single authoritative flow progress bar (spec 014): 5-segment
//  stepped mode for create (filled = accent, rest = border-level
//  neutral) and single partially-filled track for login (contract §5).
//  Decorative — the step caption / headline carry meaning for a11y.
//

import SwiftUI

struct StepProgress: View {
    enum Mode {
        /// `current` is 1-based; segments `< current` render filled.
        case stepped(current: Int, total: Int)
        /// Single track filled to `fill` (0…1).
        case single(fill: CGFloat)
    }

    @Environment(\.theme) private var theme
    let mode: Mode

    var body: some View {
        Group {
            switch mode {
            case .stepped(let current, let total):
                HStack(spacing: FlowGeometry.barGap) {
                    ForEach(0..<total, id: \.self) { index in
                        Capsule()
                            .fill(index < current ? theme.accentBase : theme.borderBase)
                            .frame(maxWidth: .infinity)
                    }
                }
            case .single(let fill):
                GeometryReader { proxy in
                    ZStack(alignment: .leading) {
                        Capsule().fill(theme.borderBase)
                        Capsule()
                            .fill(theme.accentBase)
                            .frame(width: proxy.size.width * min(max(fill, 0), 1))
                    }
                }
            }
        }
        .frame(height: FlowGeometry.barHeight)
        .accessibilityHidden(true)
    }
}

#Preview("Step progress") {
    VStack(spacing: Tokens.Space.s24) {
        StepProgress(mode: .stepped(current: 1, total: 5))
        StepProgress(mode: .stepped(current: 3, total: 5))
        StepProgress(mode: .stepped(current: 5, total: 5))
        StepProgress(mode: .single(fill: FlowGeometry.loginBarFill))
    }
    .padding(Tokens.Space.s24)
    .themed(.light)
}

#Preview("Step progress dark") {
    VStack(spacing: Tokens.Space.s24) {
        StepProgress(mode: .stepped(current: 1, total: 5))
        StepProgress(mode: .stepped(current: 5, total: 5))
        StepProgress(mode: .single(fill: FlowGeometry.loginBarFill))
    }
    .padding(Tokens.Space.s24)
    .background(Tokens.dark.bgRaised.color)
    .themed(.dark)
}

File: app-desktop/vela-wallet/src/ui/step_progress.rs
//! Progress bars for the flow Working states (spec 014): the create flow's
//! 5-segment stepped bar and the login flow's single partially-filled track —
//! one authority for both modes.

use crate::theme::{STEP_BAR_GAP, STEP_BAR_H, Theme};
use gpui::{Div, ParentElement, Styled, div, px, relative};

/// `step` is 1-based: segments `< step` render filled (mock A4: step 1 fills
/// the first segment).
pub fn step_progress(theme: &Theme, step: u8, total: u8) -> Div {
    let mut row = div().flex().w_full().gap(px(STEP_BAR_GAP));
    for ix in 0..total {
        let fill = if ix < step {
            theme.accent
        } else {
            theme.divider
        };
        row = row.child(
            div()
                .flex_1()
                .min_w(px(0.))
                .h(px(STEP_BAR_H))
                .rounded_full()
                .bg(fill),
        );
    }
    row
}

/// The login waiting bar: one full-width track, `fraction` filled accent.
pub fn login_progress(theme: &Theme, fraction: f32) -> Div {
    div()
        .w_full()
        .h(px(STEP_BAR_H))
        .rounded_full()
        .bg(theme.divider)
        .child(
            div()
                .w(relative(fraction))
                .h_full()
                .rounded_full()
                .bg(theme.accent),
        )
}
